src/build_backend.py

```diff
@@ -0,0 +1,119 @@
+import argparse
+import hashlib
+import os
+import shutil
+import subprocess
+import tarfile
+import zipfile
+
+from typing import Dict, Iterable, Optional
+
+from cockpit import __version__
+
+PACKAGE = f'cockpit-{__version__}'
+TAG = 'py3-none-any'
+
+
+def find_sources(srcpkg: bool) -> Iterable[str]:
+    try:
+        subprocess.check_call(['modules/checkout'], stdout=2)
+    except FileNotFoundError:
+        pass
+
+    if srcpkg:
+        yield from {
+            'pyproject.toml',
+            'src/build_backend.py',
+        }
+
+    for path, _dirs, files in os.walk('src', followlinks=True):
+        if '__init__.py' in files:
+            yield from [os.path.join(path, file) for file in files]
+
+
+def copy_sources(distdir: str) -> None:
+    for source in find_sources(srcpkg=True):
+        destination = os.path.join(distdir, source)
+        os.makedirs(os.path.dirname(destination), exist_ok=True)
+        shutil.copy(source, destination)
+
+
+def build_sdist(sdist_directory: str,
+                config_settings: Optional[Dict[str, object]] = None) -> str:
+    del config_settings
+    sdist_filename = f'{PACKAGE}.tar.gz'
+    with tarfile.open(f'{sdist_directory}/{sdist_filename}', 'w:gz', dereference=True) as sdist:
+        for filename in find_sources(srcpkg=True):
+            sdist.add(filename, arcname=f'{PACKAGE}/{filename}', )
+    return sdist_filename
+
+
+def build_wheel(wheel_directory: str,
```

Comment:
I'm curious: Now that we have our own wheel build anyway, is there any chance that this can be taught about libexecdir, instead of doing the `mv` dance later on?

Reply:
Yes and no.  The wheel spec (PEP 491) actually envisions exactly this: https://peps.python.org/pep-0491/#install-paths

> In addition to the distutils install paths, wheel now includes the listed categories based on GNU autotools. This expanded scheme should help installers to implement system policy, but installers may root each category at any location.
>
> A UNIX install scheme might map the categories to their installation paths like this:
> ```
> ...
>     'libexecdir': '$eprefix/libexec',
> ...
> ```

But unfortunately pip doesn't currently implement this, and complains loudly if you try to install a wheel that contains something in the libexecdir scheme :(

Reply:
There's actually a workaround there: we could dist the script as pure "data", which gets installed relative to the toplevel prefix.  In that case, though, we would hardcode "libexecdir", which would be wrong on Debian, for example.

And in any case, we couldn't take advantage of the automatic script generation — we'd have to include the scripts ourselves, complete with all the usual problems about interpreter path, etc.

There's one more aspect: files installed in this way wouldn't be marked executable.

So in general, I'd say this is probably not going to work any time soon...

Reply:
ack, thanks!